var/spack/repos/builtin/packages/dealii/package.py

```diff
@@ -170,6 +170,10 @@ class Dealii(CMakePackage, CudaPackage):
     depends_on('graphviz',         when='+doc')
     depends_on('ginkgo',           when='@9.1:+ginkgo')
     depends_on('gmsh+tetgen+netgen+oce', when='@9.0:+gmsh', type=('build', 'run'))
+    # TODO: the next line fixes concretization with flex: the scotch package,
+    # pulled in through a gmsh dependency, requires a version that conflicts
+    # with that required by one of the other dependencies.
+    depends_on('gmsh~mmg',         when='+gmsh')
```

Comment:
Are we talking about conflicting pinned versions (i.e. one package requires exactly vX and another exactly vY) or is this a limitation due to the original concretizer?

Reply:
This is the error that I was seeing before I added this line (but also before I adjusted GMSH's package):
```
$./spack spec dealii
Input spec
--------------------------------
dealii

Concretized
--------------------------------
==> Error: An unsatisfiable version constraint has been detected for spec:

    flex@2.6.3%apple-clang@12.0.0+lex~nls arch=darwin-catalina-ivybridge
        ^bison@3.7.6%apple-clang@12.0.0 arch=darwin-catalina-ivybridge
            ^diffutils@3.7%apple-clang@12.0.0 arch=darwin-catalina-ivybridge
                ^libiconv@1.16%apple-clang@12.0.0 arch=darwin-catalina-ivybridge
            ^m4@1.4.19%apple-clang@12.0.0+sigsegv arch=darwin-catalina-ivybridge
                ^libsigsegv@2.13%apple-clang@12.0.0 arch=darwin-catalina-ivybridge
            ^perl@5.34.0%apple-clang@12.0.0+cpanm+shared+threads arch=darwin-catalina-ivybridge
                ^berkeley-db@18.1.40%apple-clang@12.0.0+cxx~docs+stl arch=darwin-catalina-ivybridge
                ^bzip2@1.0.8%apple-clang@12.0.0~debug~pic+shared arch=darwin-catalina-ivybridge
                ^gdbm@1.19%apple-clang@12.0.0 arch=darwin-catalina-ivybridge
                    ^readline@8.1%apple-clang@12.0.0 arch=darwin-catalina-ivybridge
                        ^ncurses@6.2%apple-clang@12.0.0~symlinks+termlib abi=none arch=darwin-catalina-ivybridge
                            ^pkgconf@1.7.4%apple-clang@12.0.0 arch=darwin-catalina-ivybridge
                ^zlib@1.2.11%apple-clang@12.0.0+optimize+pic+shared arch=darwin-catalina-ivybridge
        ^findutils@4.8.0%apple-clang@12.0.0 arch=darwin-catalina-ivybridge


while trying to concretize the partial spec:

    scotch


scotch requires flex version :2.6.1,2.6.4:, but spec asked for 2.6.3
```

But I've now discovered that this must have been brought in by GMSH's `depends_on(gl2ps)`, because now removing that dependency clears this up. So it looks like I can in fact remove this line in the deal.II package.

Reply:
BTW. Thanks for querying this, @alalazo, and for the review.